sklearn/linear_model/tests/test_sag.py

```diff
@@ -371,7 +371,7 @@ def test_sag_regressor_computed_correctly():
     n_samples = 40
     max_iter = 50
     tol = .000001
-    fit_intercept = True
+    fit_intercept = False
```

Comment:
why do you need this @jeromedockes ? SAG fails to get a good intercept?
can you try initializing the intercept to np.mean(y_train) instead of 0 (if done like this now)?

Reply:
At the moment it does fail to fit a good intercept (with the default `tol` and
`n_iter`)

the intercept is indeed initialized with zeros:

https://github.com/scikit-learn/scikit-learn/blob/c315bf931466655191b005f9e695c0d0a1c14219/sklearn/linear_model/ridge.py#L486

but initializing it with the mean of `y` probably won't change much since this
mean is `0`: `y` is always assumed to be dense and centered in preprocessing.
The mean of `X` is what causes the intercept to be nonzero.

Reply:
I was thinking that we could revert to only allowing `sparse_cg` for sparse data
in this PR to quickly fix the bug and unlock #13246 , and then see if support
for sparse data can be added to the `sag` solver in a separate PR. WDYT?

Reply:
I am +1 for this path. I think that we should ensure that we give proper results. We can investigate later how to fix SAG for this case. @agramfort WDYT?

Reply:
BTW one of the reasons the intercept takes many iterations to converge can be this decay

https://github.com/scikit-learn/scikit-learn/blob/4a6264db68b28a2e65efdecc459233911c9aee95/sklearn/linear_model/base.py#L43

https://github.com/scikit-learn/scikit-learn/blob/4a6264db68b28a2e65efdecc459233911c9aee95/sklearn/linear_model/base.py#L97

https://github.com/scikit-learn/scikit-learn/blob/4a6264db68b28a2e65efdecc459233911c9aee95/sklearn/linear_model/sag.py#L299